docs: AuthZ (#4186)
apollographql/router: docs/source/configuration/authorization.mdx

> ### `@policy`

Reviewer: Is it too late to change the name? 😅 Seems like `requiresPolicies` would be more consistent with the other `requiresScopes` directive.

The `@policy` directive marks fields and types as restricted based on authorization policies evaluated in a [Rhai script](../customizations/rhai/) or [coprocessor](../customizations/coprocessor). This enables custom authorization validation beyond authentication and scopes.

> The `@policy` directive includes a `policies` argument that defines an array of the required policies. The following example shows a policy that requires all roles from the claims object to contain the string `"support"`:

Reviewer: I think it'd be better here to use a generic policy string, not Rhai code (I'm not sure we should be suggesting Rhai code in general, more on that later). This paragraph & example make it seem like the default behavior of policies is to execute some code, which isn't the case.

Author: cc @Geal, for this and other comments from Dylan

```graphql
@policy(policies: [[]"claims[`roles`].contains(`support`)"]])
```

At the [`RouterService` level](../customizations/overview#the-request-lifecycle), the Apollo Router extracts the list of policies relevant to a request from the schema and then stores them in the request's context in `apollo_authorization::policies::required` as a map `policy -> null|true|false`.

> At the `SupergraphService` level, you need to provide a Rhai script or a coprocessor that evaluates the map. If the policy is validated, set its value to `true`, otherwise set it to `false`. Afterwards, the router filters the requests' types and fields to only those where the policy is `true`.

Reviewer 2: ```suggestion
At the `SupergraphService` level, you must provide a Rhai script or coprocessor to evaluate the map. If the policy is validated, the script or coprocessor should set its value to `true` or otherwise set it to `false`. Afterward, the router filters the requests' types and fields to only those where the policy is `true`.
```

Reviewer: What happens if the policy is still `null` (e.g., because it's not implemented in the coprocessor yet)?

> If no field of a subgraph query passes its authorization policies, the router stops further processing of the query and precludes unauthorized subgraph requests. This efficiency gain is a key benefit of the `@policy` directive and other authorization directives.

Reviewer 2: ```suggestion
If no field of a subgraph query passes its authorization policies, the router stops further processing of the query and precludes unauthorized subgraph requests. This efficiency gain is a key benefit of the `@policy` and other authorization directives.
```

>     import: [..., "@policy"])

Reviewer: I assume this will need to be fed 2.6 since 2.5 is out and doesn't support this yet

```

The `@policy` directive is defined as follows:

```graphql
scalar federation__Policy
directive @policy(scopes: [[federation__Policy!]!]!) on OBJECT | FIELD_DEFINITION | INTERFACE | SCALAR | ENUM
```

Using the `@policy` directive requires a [Supergraph plugin](../customizations/overview) to evaluate the authorization policies. You can do this with a [Rhai script](../customizations/rhai/) or [coprocessor](../customizations/coprocessor). Refer to the following [Rhai script](#usage-with-a-rhai-script) and [coprocessor](#usage-with-a-coprocessor) examples for more information. (Although a [native plugin](../customizations/native) can also evaluate authorization policies, we don't recommend using it.)

> #### Combining policies with `AND`/`OR` logic

Reviewer: Should we combine this with the `@requiresScopes` description of AND/OR? I believe it works the same way, but maybe it's better to keep the definition separate, unsure.

Authorization validation uses **AND** logic between the elements in the inner-level `policies` array, where a request must include _all_ elements in the inner-level `policies` array to resolve the associated field or type. For the following example, a request would need `policy1` **AND** `policy2` **AND** `policy3` to be authorized:

```graphql
@requiresScopes(policies: [["policy1", "policy2", "policy3"]])
```

Alternatively, to introduce **OR** logic you can use nested arrays. For the following example, a request would need `policy1` **OR** `policy2` **OR** `policy3` to be authorized:

```graphql

> @requiresScopes(policies: [["policy1"], ["policy2"], ["policy3"]])

Reviewer: ```suggestion
@policy(policies: [["policy1"], ["policy2"], ["policy3"]])
```

```

You can nest arrays and elements as needed to achieve your desired logic. For the following example, its syntax requires requests to have either (`policy1` **AND** `policy2`) **OR** just `policy3` to be authorized:

```graphql
@policy(policies: [["policy", "policy2"], ["policy3"]])
```

##### Usage with a Rhai script

> The `policies` argument contains a list of strings with no formatting constraints, meaning you can use them to store Rhai code.

Reviewer: I don't think we should recommend that folks do `eval` of arbitrary Rhai code from the subgraphs. There are a couple of problems:

1. You're suddenly giving developers who have access to subgraphs access to run code at the router level, which could let them get data they shouldn't have somehow? Maybe not the hugest issue, but worries me.
2. If someone writes invalid Rhai code in _any_ subgraph, they could break **every** operation which tests the policy—it wouldn't just cause that new field to get omitted, it'd cause entire requests to be aborted with an error.

      context: true
```

> A coprocessor can then receive a request with this format:

Reviewer: The coprocessor example feels like one that could use `requiresScopes`. Can we reframe it to make it more obvious why you'd use `policy` in some cases and when you _should_ use `requiresScopes` instead? This might be a bigger point we want to make anyway, why does `@requiresScopes` exist if `@policy` is more powerful?

## Configuration options

The behavior of the authorization plugin can be modified with various options.

### reject_unauthorized

> The `reject_unauthorized` option configures whether to reject an entire query if any authorization directive failed, where any part of it is filtered by authorization directives. When enabled, a response contains the list of paths that are affected.

Reviewer 2: ```suggestion
The `reject_unauthorized` option configures whether to reject an entire query if any authorization directive failed, or any part of the query was filtered by authorization directives. When enabled, a response contains the list of paths that are affected.
```

      log: false # default: true
```

<Note>

The `log` option should be disabled if filtering parts of queries according to the client's rights is approved as normal operation by platform operators.

</Note>

#### response

> You can configure `response` to set where to place filtered paths in a GraphQL response:

Reviewer 2: ```suggestion
You can configure `response` to define what part of the GraphQL response should include filtered paths:
```
